addon/services/prefetch.js

```diff
@@ -28,11 +29,20 @@ if (gte('3.6.0')) {
                 let { route, fullParams } = changeSet.for[i];
                 if (seenRoutes.has(route)) continue;
 
-                  route._prefetched = new RSVP.Promise(r => {
-                    return r(route.prefetch(fullParams, transition))
+                if (isThenable(route)) {
+                  // Ensure we use the expected fullParams when the promise then
+                  // handler is executed for lazy loaded routes.
+                  RSVP.resolve(fullParams).then(capturedFullParams => {
+                    route.then(resolvedRoute => {
+                      route._prefetched = RSVP.Promise.resolve(resolvedRoute.prefetch(capturedFullParams, transition));
```

Comment:
I am thinking also it probably makes sense to check isDestroying and isAborted in this callback, but I'll wait for your next review to see what your thoughts are.